PLANNER-2435: Kogito task assigning phase1 blog post (#358)
kiegroup/optaplanner-website: content/blog/2021/06/15/LetsOptaPlanYourJbpmTasksPart2.adoc

= Let's OptaPlan your jBPM tasks (part 2) - Moving to Kogito
wmedvede
2021-06-15
:page-interpolate: true
:jbake-type: post
:jbake-tags: feature, task assignment, integration
:jbake-social_media_share_image: CreditDisputeProcessDiagram.png

> In the previous blog post link:../../../2020/07/03/LetsOptaPlanYourJbpmTasksPart1.html[Let's OptaPlan your jBPM tasks (part 1)] we showed that OptaPlanner can be used for assigning the human tasks produced by a set of business processes to users according to an optimized plan.

Reviewer: Can we make the first paragraph more answer the question "What is the benefit for me reading this?" What is the value proposition? Why should I care?

Referencing the previous blog is good, but I 'd leave that to the second paragraph, if you want them to read that first to better understand the problem. Readers are lazy, the TLDR of the problem/value should be in the first paragraph.

Readers might not now what Kogito is or what the benefit of kogito is.

Maybe something in the lines of
"In BPMN workflows in the cloud, we can automatically assign tasks to users by combining Kogito processes with OptaPlanner optimization. Furthermore, we can optimize those assignments, to ensure every user has the right skills for the tasks and ..."

This cycle is repeated over and over.

== Process execution

Let's now execute the process in the https://kogito.kie.org/[Kogito] ecosystem.

image::CreditDisputeProcessDiagramWithTaskInputs.png[]

The following image shows how a new process instance is created by using the Kogito generated endpoints:

image::Start_CreditDispute_1111_MASTER_Spanish.png[]

Analogous the link:../../../2020/07/03/LetsOptaPlanYourJbpmTasksPart1.html[Let's OptaPlan your jBPM tasks (part 1)] case, this information combined with the user's configuration is used to establish which of the available users are best suited for each task.

This time, the following user's configuration is used:

[cols="10%,40%,50%" frame="all", options="header"]
|===
|User
|Belongs the group
|Has skills in

|katy | CreditAnalyst | CITI and MASTER cards.
|emily | CreditAnalyst | VISA and AMERICAN_EXPRESS cards.
|bob | ClientRelations | Spanish and English languages.
|frank | ClientRelations | Chinese language.

|===

When the process is started a new instance of the ResolveDispute task is created and automatically assigned:

image::ResovleDispute_1111_MASTER_Spanish_katy.png[]

> We can see that the task requires "a CreditAnalyst user with skills in the MASTER credit card". This is why it was assigned to katy.

Reviewer: katy with a capital?

Writing docs guidelines is use active tense, which works well in blogs too:
"we can see that the task requires" => "The task requires ..."

As soon this task is completed, a new instance of the Notify Customer task is created.

The following image shows the new instance information:

image::NotifyCustomer_1111_MASTER_Spanish_bob.png[]

> We can see that the Spanish language was selected as the "Preferred language for communications", which establishes that:

Reviewer: ```suggestion
Notice that the Spanish language was selected as the "Preferred language for communications", which establishes that:
```

"The task requires a ClientRelations user with skills in the Spanish language", and this is why this new task is assigned to bob.

As in the previous blog post the example above is quite linear, only four users are defined, and the possible assignments are direct. This was done intentionally to show how the assignment decisions are made accordingly with the configured users, groups, and the tasks business data.

== Conclusions

OptaPlanner can be used in different ways, ranging from stand-alone CLI to complex web applications like vehicle routing, Kie Server based applications, https://quarkus.io/guides/optaplanner[Quarkus] applications, and also as part of any https://kogito.kie.org/[Kogito] application.

> In general, to integrate OptaPlanner in your https://kogito.kie.org/[Kogito] application is straightforward, you instruct OptaPlanner with the model, and the Solver configuration to use, and that’s all ;), OptaPlanner link:../../../../compatibility/quarkus.html[Quarkus] or link:../../../../compatibility/springBoot.html[SpringBoot] extensions will manage all the complexity of the Solver instantiation and management.

Reviewer: "is straightforward, you instruct" => "is straightforward: you instruct"
